---
title: "Methylation Array Analysis"
author: "Jihwan Lim & Inkyun Park"
date: "2022-12-21"
output: html_document
---
# Methylation Array Analysis

A methylation array data set was analysed to assess methylation changes in tumor tissue versus normal lung tissue. The data was collected from lung cancer patients and normal people in Norway. DNA from patients and people were analysed with a Illumina Infinium HumanMethylation450 BeadChip.(GSE40419)

## Load in necessary packages
```{r warning=FALSE, results='hide', message=FALSE}
library(tidyverse)
library(lumi)
library(wateRmelon)
library(ChAMPdata)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(org.Hs.eg.db)
```

## Load annotation data

```{r}
infinium_annotation <- t(read.table("/Users/jihwanlim/Desktop/Ghent/Msc/Applied_High-throughput_Analysis/Project/Infinium_r/GSE66836_series_matrix.txt",sep="\t",fill=T))
infinium_annotation <- data.frame(ID = rownames(infinium_annotation), infinium_annotation)
infinium_annotation[1,1] <- "ID"
colnames(infinium_annotation) <- infinium_annotation[1,]
infinium_annotation <- infinium_annotation[-1,]
rownames(infinium_annotation) <- 1:nrow(infinium_annotation)
head(infinium_annotation[,c(2,3,9,10,12,13,14)])
```

## select data 

### Get specific data that we target

```{r}
# Pick necessary colums for choosing
annot <- infinium_annotation[c("!Sample_title","!Sample_geo_accession", "!Sample_characteristics_ch1", "!Sample_characteristics_ch2", "!Sample_characteristics_ch3", "!Sample_characteristics_ch4", "!Sample_characteristics_ch5", "!Sample_source_name_ch1", "!Sample_description1", "!Sample_description2")]

# Change the name of elements to do remove unnecessary data
annot$`!Sample_characteristics_ch1` <- gsub('tissue: ','', annot$`!Sample_characteristics_ch1`)
annot$`!Sample_description2` <- gsub('Sentrix_Position: ','', annot$`!Sample_description2`)
annot$`!Sample_description1` <- gsub('Sentrix_ID: ','', annot$`!Sample_description1`)
head(annot, 10)
```

```{r}
# Pick necessary columns for choosing
annot_sel <- infinium_annotation[c("!Sample_title","!Sample_geo_accession", "!Sample_characteristics_ch1", "!Sample_characteristics_ch2", "!Sample_characteristics_ch3", "!Sample_characteristics_ch4", "!Sample_characteristics_ch5", "!Sample_source_name_ch1", "!Sample_description1", "!Sample_description2")]

# Change the name of elements to do remove unnecessary data
annot_sel$`!Sample_characteristics_ch1` <- gsub('tissue: ','', annot_sel$`!Sample_characteristics_ch1`)
annot_sel$`!Sample_description2` <- gsub('Sentrix_Position: ','', annot_sel$`!Sample_description2`)
annot_sel$`!Sample_description1` <- gsub('Sentrix_ID: ','', annot_sel$`!Sample_description1`)


colnames(annot_sel) <- c("Sample_title","Geo_accession","Tissue", "Stage", "p53_status", "EGFR_status", "KRAS_status", "character","Sentrix_ID", "Sentrix_Position")

annot_sel$Stage <- gsub("Stage: ", "", annot_sel$Stage)
annot_sel$p53_status <- gsub("p53 status: ", "", annot_sel$p53_status)
annot_sel$EGFR_status <- gsub("egfr status: ", "", annot_sel$EGFR_status) 
annot_sel$KRAS_status <- gsub("kras status: ", "", annot_sel$KRAS_status)

head(annot_sel, 10)
```

### filtration for annotation

```{R}
des1 <- annot$`!Sample_geo_accession`
des2 <- annot$`!Sample_description1`
des3 <- annot$`!Sample_description2`

# how sample name looks like
des_final <- paste(des1,des2,des3,sep = "_")

annot$marker <- des_final
annot_sel$marker <- des_final
```

### Get annoation of stage 3 tumor samples with mutated EGFR

```{r}
Tumor <- annot_sel[grep("Mutated", annot_sel$EGFR_status),]
Tumor <- Tumor[grep("WildType", Tumor$p53_status),]
Tumor <- Tumor[grep("WildType", Tumor$KRAS_status),]
Tumor3 <- Tumor[grep(3 ,Tumor$Stage),]
Tumor3
```

### Get annotation of normal samples

```{r}
Normal <- annot_sel[grep("Normal", annot_sel$Tissue),]

# Find normal samples that have same Sentrix position with tumor 3 
Normal <- Normal[Normal$Sentrix_Position %in% Tumor3$Sentrix_Position,]
Normal
```

we have 19 normal samples and 164 tumor samples. Since there are lots of factors to be considered like stage, mutation of EGFR, KRAS, or TP53 genes, etc, we decided to use stage 3 with mutated EGFR samples (WT with KRAS and TP53 genes) for tumor samples. This can be further related to RNA-seq data which has stage 3 tumor tissue. Then we randomly chose 3 normal samples which have same sentrix position among 5 of samples.

## Load the Infinium data
```{r warning=FALSE, results='hide', message=FALSE}
# Load EPIC data
infdata <- readEPIC("/Users/jihwanlim/Desktop/Ghent/Msc/Applied_High-throughput_Analysis/Project/Infinium_r/tumor3/")
# Since there are 183 samples in raw file, we just made a new folder that only contains that we only chose before
```

## Take new annotation table that only contain necessary data
```{r}
# we already make marker column 
annot <- annot %>% filter(annot$marker %in% sampleNames(infdata))
```

## Have a look at the data and annotation
```{r}
print(infdata)
print(dim(infdata))
print(annot)
print(sum(is.na(exprs(infdata))))

# betas function retrieve beta value (=methylation percentage)
print(head(betas(infdata)))

# exprs function retrieve M-value
print(head(exprs(infdata)))
```

## Preprocessing the data

```{r}
# Remove all NA value both in M-value and Methylation percentage
infdata <- infdata[rowSums(is.na(exprs(infdata))) == 0,]
head(exprs(infdata))
```

## Explore preprocessed data

```{r}
head(exprs(infdata))
```

## Change samplNAMES to somthing more comprehensible

```{r}
sampleNames(infdata) <- paste(annot[,2], annot[,3], sep = "_")
```

## Remove probes for which calling p-value insufficient

```{r}
infdata_filt <- pfilter(infdata)
```

## Comparison of average methylation between control and cancer samples

```{r}
boxplot(betas(infdata_filt), las=2)
```


```{r}
control <- (infdata_filt[,grep("Normal",annot[,3])])
cancer <- (infdata_filt[,grep("Tumor",annot[,3])])

meth_mean_CAF <- rep(0,ncol(cancer))
meth_mean_NAF <- rep(0,ncol(control))

for (i in 1:ncol(cancer)){
  meth_mean_CAF[i] <- mean(betas(cancer[,i]))
}

for (i in 1:ncol(control)){
  meth_mean_NAF[i] <- mean(betas(control[,i]))
}

meth_mean_CAF
meth_mean_NAF

t_test_res <- t.test(meth_mean_NAF, meth_mean_CAF, var.equal = F)
t_test_res
```

## Normalization and QC

```{r}
infdata_norm <- dasen(infdata_filt)
head(infdata_norm)
```

### Mkake methylumi objects to check density and color bias adjustment

```{r}
infdataM_norm <- as(infdata_norm, "MethyLumiM")
infdataM <- as(infdata_filt, "MethyLumiM")
```

### Make QC plot

```{r}
par(mfrow = c(2,2))
plotColorBias1D(infdataM, channel="both", main="before")
plotColorBias1D(infdataM_norm, channel="both",main="after")
density(infdataM, xlab="M-value", main="before", legend =F)
density(infdataM_norm, xlab="M-value", main="after", legend = F)
```

## Differential expression analysis

```{r}
# Define design matrix
des <- factor(as.character(annot[,3]))
design <- model.matrix(~0 + des)
colnames(design) <- c("Tumor","Normal")
fit <- lmFit(infdataM_norm, design)
```

```{r}
# Fitting the model
cont.matrix <- makeContrasts(NvsS=Tumor-Normal,levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2)

# Getting top genes
topTable(fit2, coef=1, adjust="BH")
results <- decideTests(fit2)
vennDiagram(results)
summary(results)
```

## DE results

```{r}
LIMMAout <- topTable(fit2, adjust="BH", number=nrow(exprs(infdataM)))
head(LIMMAout, 10)
```

### Volcano plot

```{r}
# There is few signifcant genes with threshold 0.05, so 0.15 is used instead
volcanoplot(fit2, col = as.factor(LIMMAout$adj.P.Val < 0.15), style = "p-value")
```

### MA plot

```{r}
plot(LIMMAout$AveExpr, LIMMAout$logFC,
     col = as.factor(LIMMAout$adj.P.Val < 0.05), pch = 20, cex = 0.50,
     xlab = "Average Intensities", ylab = "logFC")
```

## Functional annotation of limma results

### Load annotation and sort alphabetically on probe name

```{r}
data("probe.features")
annotation_MA <- probe.features
print(head(annotation_MA))
annotation_MA <- annotation_MA[sort(rownames(annotation_MA),index.return = T)$ix,]
```

### Check if all probes are present in both sets

```{r}
dim(LIMMAout)
sum(LIMMAout$Probe_ID%in%rownames(annotation_MA))
sum(rownames(annotation_MA)%in%LIMMAout$Probe_ID)
# Also check the reverse so no duplicate rows are present in annotation
```
### Since more probes are present in the annotation file, remove unnecessary probes

```{r}
annotation_MA <- annotation_MA[rownames(annotation_MA)%in%LIMMAout$Probe_ID,]
```

### Sort LIMMA output alphabetically on probe name

```{r}
LIMMAout_sorted <- LIMMAout[sort(LIMMAout$Probe_ID,index.return=T)$ix,]
```

### Add gene names to LIMMA output
```{r}
LIMMAout_sorted$Gene <- annotation_MA$gene
LIMMAout_sorted$Feature <- annotation_MA$feature
LIMMAout_sorted$Chrom <- annotation_MA$CHR
LIMMAout_sorted$Pos <- annotation_MA$MAPINFO
LIMMAout_sorted$Chrom <- as.character(LIMMAout_sorted$Chrom)
LIMMAout_sorted$Gene <- as.character(LIMMAout_sorted$Gene) 
LIMMAout_sorted$Feature <- as.character(LIMMAout_sorted$Feature)
```

## Quantification of absolute methylation differences

### Add gene names to LIMMA output
```{r}
LIMMAout_sorted$Tumor_meth <- rowMeans(betas(infdata)[rownames(infdata)%in%
                                                        LIMMAout_sorted$Probe_ID,annot$`!Sample_characteristics_ch1`=="Tumor"]) 
LIMMAout_sorted$Control_meth <- rowMeans(betas(infdata)[rownames(infdata)%in%
                                                          LIMMAout_sorted$Probe_ID,annot$`!Sample_characteristics_ch1`=="Normal"])
LIMMAout_sorted$Abs_diff_meth <- abs(rowMeans(betas(infdata)[rownames(infdata)%in%
                                                               LIMMAout_sorted$Probe_ID,annot$`!Sample_characteristics_ch1`=="Tumor"]) - 
                                       rowMeans(betas(infdata)[rownames(infdata)
                                                               %in%LIMMAout_sorted$Probe_ID, annot$`!Sample_characteristics_ch1`=="Normal"]))
```

## Resort results

```{r}
LIMMAout_annot <- LIMMAout_sorted[sort(LIMMAout_sorted$P.Value,index.return=T)$ix, c(1,12,13,10,11,4,7,8,5,14,15,16)]
# Sort on p-values to prevent errors in sorting due to equal FDR values
```

## Interpretation results

### Select CpGs in genic regions

```{r}
sum(LIMMAout_annot$adj.P.Val<0.05) 
sum(LIMMAout_annot$adj.P.Val[LIMMAout_annot$Gene!=""]<0.05)

LIMMAout_annot_gene <- LIMMAout_annot[LIMMAout_annot$Gene!="",]
```

### Check genic results

```{r}
head(LIMMAout_annot_gene[c(4,5,6,8,10,11,12)])

topgenes_genic <- unique(LIMMAout_annot_gene$Gene[1:10])

for (i in 1:length(topgenes_genic)){
  LIMMAout_subset <- LIMMAout_annot_gene[(LIMMAout_annot_gene$Gene==topgenes_genic [i]) &
                                           (LIMMAout_annot_gene$adj.P.Val<0.05) &
                                           (abs(LIMMAout_annot_gene$logFC)>2),]
  print(LIMMAout_subset[sort(LIMMAout_subset$Pos,index.return=T)$ix,c (4,5,6,8,10,11,12)])
  }
```

### Select CpGs in promoter regions

```{r}
LIMMAout_annot_prom <- LIMMAout_annot_gene[grepl("TSS",LIMMAout_annot_gene$Feature) | (LIMMAout_annot_gene$Feature=="1stExon"),]

head(LIMMAout_annot_prom)
## Look for multiple CpG in promoter regions undergoing similar methylation differences


topgenes_prom <- unique(LIMMAout_annot_prom$Gene[1:10]) 
```

```{r}
for (i in 1:length(topgenes_prom)){
  LIMMAout_subset <- LIMMAout_annot_prom[(LIMMAout_annot_prom$Gene == topgenes_prom[i]) & (LIMMAout_annot_prom$adj.P.Val < 0.10),]
  if (nrow(LIMMAout_subset) > 1) {print(LIMMAout_subset[sort(LIMMAout_subset$Pos, index.return =T)$ix, c(4,5,6,8,10,11,12)])
    }
 }
```

## Gene Set Analysis

Goana uses Entrez gene identifiers, we used to convert our gene symbols to entrez ids. For thus purpose we use the org.Hs.eg.db package.

```{r}
LIMMAout_filtered <- LIMMAout_annot[LIMMAout_annot$adj.P.Val < 0.05,]

EntrezIDs <- mapIds(org.Hs.eg.db, LIMMAout_filtered$Gene, "ENTREZID", "SYMBOL")
```

### subset for non duplicated and mapped genes

```{r}
EntrezIDs <- EntrezIDs[!(duplicated(EntrezIDs) | is.na(EntrezIDs))]
head(EntrezIDs)
```

### Make table for comaprison with other methods
```{r}
LIMMAout_filtered$EntrezIDs <- EntrezIDs[match(LIMMAout_filtered$Gene, names(EntrezIDs))]
```

### Overexpression analysis with goana

```{r}
goanaOUT <- goana(de=unlist(EntrezIDs), species = "Hs", trend = T)
```

### FDR multiple

```{r}
goanaOUT <- goanaOUT[order(goanaOUT$P.DE, decreasing = F),]
goanaOUT$FDR.DE <- p.adjust(goanaOUT$P.DE, method = "BH")

topGOcpg <- topGO(goanaOUT, ontology = "BP", number = 50)
head(topGOcpg)
```

```{r}
head(topGOcpg[order(topGOcpg$N),])
```

## Write data for comaprison of results

```{r}
write.table(unlist(EntrezIDs), sep = "\t", file = "EntrezIDs_CpG_results.txt")
```

```{r}
CpG_GSA_res <- topGO(goanaOUT, ontology = "BP", number = 100)
write.table(CpG_GSA_res, sep="\t", file = "CpG_GSA_results.txt")
```
